blog/optimizing-clickhouse-intel-high-core.md

```diff
@@ -0,0 +1,817 @@
+# up to 64x larger than the requested allocation size
+
+# After: Optimized jemalloc configuration
+JEMALLOC_CONFIG_MALLOC_CONF = "oversize_threshold:0,muzzy_decay_ms:0,dirty_decay_ms:5000,lg_extent_max_active_fit:8"
+# Key parameter explained:
+# lg_extent_max_active_fit:8 - Allows memory reuse from extents up to 256x larger 
+# than the requested allocation size (2^8 = 256x vs default 2^6 = 64x)
+# The 256x limit matches ClickHouse's two-level hash table structure (256 buckets)
+# This enables efficient reuse of merged hash table memory blocks
+```
+
+**Performance Impact Analysis**:
+
+The improvement comes from:
+1. **Better Memory Reuse**: Deallocated memory becomes available for reuse
+2. **Reduced Page Faults**: Fewer new page allocations needed
+3. **Lower Resident Memory**: Physical memory usage remains bounded
+4. **Improved Cache Locality**: Better memory layout reduces cache misses
+
+**Results**:
+- **ClickBench Q35**: 96.1% performance improvement (1.961x queries per second)
+- **Memory Usage**: 45.4% reduction in VmRSS (resident memory)
+- **Page Faults**: 71% reduction in total page faults (0.29x vs baseline)
+- **CPU Efficiency**: 99.3% improvement in IPC (Instructions Per Cycle)
+- Eliminated memory-related query performance degradation over time
+- Improved system stability under high memory pressure
+
+**Key Insight**: Memory allocator behavior can have dramatic impacts on high core count performance. Tuning allocator parameters is often necessary for optimal scalability.
+
+### Example 2.2: AST/Analyzer Query Rewriting for Memory Reduction (PR #57853)
+
+**Problem Identified**: Q29 was memory-bound with excessive memory throughput from redundant computations in expressions like `sum(column + literal)`.
+
+**Deep Dive: Frontend Query Optimization for Memory Efficiency**
+
+This optimization demonstrates how frontend query rewriting can dramatically reduce memory pressure by eliminating redundant computations. The key insight is that many analytical queries contain patterns that can be algebraically simplified.
+
+**Understanding the Memory Bottleneck**:
+
+ClickBench Q29 contains multiple sum expressions with literals:
+```sql
+SELECT SUM(ResolutionWidth), SUM(ResolutionWidth + 1), SUM(ResolutionWidth + 2), 
+       SUM(ResolutionWidth + 3), SUM(ResolutionWidth + 4), SUM(ResolutionWidth + 5), 
+       SUM(ResolutionWidth + 6), SUM(ResolutionWidth + 7), SUM(ResolutionWidth + 8), 
+       SUM(ResolutionWidth + 9), SUM(ResolutionWidth + 10), SUM(ResolutionWidth + 11), 
+       SUM(ResolutionWidth + 12), SUM(ResolutionWidth + 13), SUM(ResolutionWidth + 14), 
+       SUM(ResolutionWidth + 15), SUM(ResolutionWidth + 16), SUM(ResolutionWidth + 17), 
+       SUM(ResolutionWidth + 18), SUM(ResolutionWidth + 19), SUM(ResolutionWidth + 20),
+       -- ... continues up to SUM(ResolutionWidth + 89)
+FROM hits;
+```
+
+The original execution plan:
+1. **Load Column**: Read ResolutionWidth values from storage repeatedly (90 times)
```

Comment:
Actually, `ResolutionWidth` is read once from storage. Only the aggregation happened repeatedly and unnecessarily.